File: login-server/src/main/java/work/fking/pangya/login/net/ClientPacketDispatcher.java
package work.fking.pangya.login.net;

import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import work.fking.pangya.login.LoginServer;
import work.fking.pangya.login.Player;

public class ClientPacketDispatcher extends SimpleChannelInboundHandler<ClientPacket> {

    private static final Logger LOGGER = LoggerFactory.getLogger(ClientPacketDispatcher.class);

    private final LoginServer loginServer;
    private final Player player;
    private final ClientPacketHandler[] handlers;

    private ClientPacketDispatcher(LoginServer loginServer, Player player, ClientPacketHandler[] handlers) {
        this.loginServer = loginServer;
        this.player = player;
        this.handlers = handlers;
    }

    public static ClientPacketDispatcher create(LoginServer gameServer, Player player, ClientPacketHandler[] handlers) {
        return new ClientPacketDispatcher(gameServer, player, handlers);
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, ClientPacket packet) {
        var packetId = packet.type().id();
        var handler = handlers[packetId];

        if (handler == null) {
            throw new IllegalStateException("Packet " + packet.type() + " has no attached handler");
        }
        var buffer = packet.buffer();

        try {
            handler.handle(loginServer, player, buffer);
        } finally {
            buffer.release();
        }
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        LOGGER.error("Exception caught in dispatcher", cause);
        ctx.disconnect();
    }
}

File: login-server/src/main/kotlin/work/fking/pangya/login/net/ClientPacketDispatcher.kt
package work.fking.pangya.login.net

import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler
import org.slf4j.LoggerFactory
import work.fking.pangya.login.LoginServer
import work.fking.pangya.login.Player

class ClientPacketDispatcher(
    private val loginServer: LoginServer,
    private val player: Player,
    private val handlers: Array<ClientPacketHandler?>
) : SimpleChannelInboundHandler<ClientPacket>() {

    companion object {
        private val LOGGER = LoggerFactory.getLogger(ClientPacketDispatcher::class.java)
    }

    override fun channelRead0(ctx: ChannelHandlerContext, packet: ClientPacket) {
        val packetId = packet.type.id()
        val handler = handlers[packetId] ?: throw IllegalStateException("Packet " + packet.type + " has no attached handler")
        val buffer = packet.buffer
        try {
            handler.handle(loginServer, player, buffer)
        } finally {
            buffer.release()
        }
    }

    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        LOGGER.error("Exception caught in dispatcher", cause)
        ctx.disconnect()
    }
}